ppx/reason_react_ppx.ml

```diff
@@ -45,16 +45,163 @@ let merlinFocus =
 let nolabel = Nolabel
 let labelled str = Labelled str
 let optional str = Optional str
+let externalDeclarations = ref []
+
+let externalExists name declarations =
+  List.exists
+    (fun decl ->
+      match decl.pstr_desc with
+      | Pstr_primitive { pval_name; _ } -> pval_name.txt = name
+      | _ -> false)
+    declarations
+
+let getLabelOrEmpty label =
+  match label with Optional str | Labelled str -> str | Nolabel -> ""
+
+let isDataProp label =
+  let labelStr = getLabelOrEmpty label in
+  String.length labelStr >= 5 && String.sub labelStr 0 5 = "data_"
+
+let transformToKebabCase name =
+  if String.length name > 5 && String.sub name 0 5 = "data_" then
+    let suffix = String.sub name 5 (String.length name - 5) in
+    let kebabSuffix = String.map (function '_' -> '-' | c -> c) suffix in
+    "data-" ^ kebabSuffix
+  else name
+
+let generateExternalName ~elementName ~props =
+  let propNames = List.map (fun (label, _) -> getLabelOrEmpty label) props in
+  let propString = String.concat "_" propNames in
+  let hash = Digest.to_hex (Digest.string propString) in
+  Printf.sprintf "makeProps_%s_%s" elementName (String.sub hash 0 8)
+
+let createMelObjAttribute ~loc =
+  {
+    attr_name = { txt = "mel.obj"; loc };
+    attr_payload = PStr [];
+    attr_loc = loc;
+  }
+
+let createMelAsAttribute ~loc jsName =
+  {
+    attr_name = { txt = "mel.as"; loc };
+    attr_payload =
+      PStr
+        [
+          Builder.pstr_eval ~loc
+            (Builder.pexp_constant ~loc (Pconst_string (jsName, loc, None)))
+            [];
+        ];
+    attr_loc = loc;
+  }
+
+let createWarningSuppressionAttribute ~loc =
+  {
+    attr_name = { txt = "warning"; loc };
+    attr_payload =
+      PStr
+        [
+          Builder.pstr_eval ~loc
+            (Builder.pexp_constant ~loc (Pconst_string ("-32", loc, None)))
+            [];
+        ];
+    attr_loc = loc;
+  }
+
+let rec buildArrowType ~loc props =
+  match props with
+  | [] ->
+      Builder.ptyp_arrow ~loc Nolabel
+        (Builder.ptyp_constr ~loc { txt = Lident "unit"; loc } [])
+        (Builder.ptyp_var ~loc "a")
+  | (label, _) :: rest ->
+      let propName = getLabelOrEmpty label in
+      let propType =
+        if propName = "children" then
+          Builder.ptyp_constr ~loc
+            { txt = Ldot (Lident "React", "element"); loc }
+            []
+        else if propName = "style" then
+          Builder.ptyp_constr ~loc
+            { txt = Ldot (Ldot (Lident "ReactDOM", "Style"), "t"); loc }
+            []
+        else if propName = "onClick" then
+          Builder.ptyp_arrow ~loc Nolabel
```

Comment:
Haven't followed the transformations yet, but this looks extremly ad-hoc for 3 props? 

I don't know if this is intentional, ai artifact or not even used, but I suspect this isn't the quality we can afford in a ppx that carries a ton of tech debt already.